### PR TITLE
Implement OnChanUpgradeInit on Controller Chain for interchain-accounts

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #5140

Spec outlined [here](https://github.com/cosmos/ibc/blob/main/spec/app/ics-027-interchain-accounts/README.md#upgrade-handshake)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go [hunk 1]
```diff
@@ -85,6 +85,11 @@ func (k Keeper) GetConnectionID(ctx sdk.Context, portID, channelID string) (stri
 	return channel.ConnectionHops[0], nil
 }
 
+// GetChannelKeeper returns the ChannelKeeper
+func (k Keeper) GetChannelKeeper() icatypes.ChannelKeeper {
+	return k.channelKeeper
+}
```
Comment (chatton) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:91]: this is a bit of a hack, `icatypes.ValidateControllerMetadata` explicitly requires a `ChannelKeeper`, this was added to be able to provide that.

File: scripts/go-lint-all.sh [hunk 2]
```diff
@@ -9,10 +9,10 @@ lint_module() {
   local root="$1"
   shift
   cd "$(dirname "$root")" &&
-    echo "linting $(grep "^module" go.mod) [$(date -Iseconds -u)]" &&
+    echo "linting $(grep "^module" go.mod) [$(date -u +"%Y-%m-%dT%H:%M:%S")]" &&
```
Comment (chatton) [inline scripts/go-lint-all.sh:12]: updated to the version on main which works on macOS

**Comment by chatton** [comment]
want to do small refactor on this, don't review yet!

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go (hunk 1 above)
Comment (chatton) [reply modules/apps/27-interchain-accounts/controller/keeper/keeper.go]: hack no longer required

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 3]
```diff
@@ -136,3 +137,42 @@ func (Keeper) OnChanCloseConfirm(
...
+	// must remain the same after the upgrade.
+	if currentMetadata.Address != metadata.Address {
+		return "", errorsmod.Wrap(icatypes.ErrInvalidAccountAddress, "address cannot be changed")
+	}
+
+	if currentMetadata.ControllerConnectionId != connectionHops[0] {
+		return "", errorsmod.Wrap(connectiontypes.ErrInvalidConnectionIdentifier, "proposed connection hop must not change")
+	}
+
+	metadataBz := icatypes.ModuleCdc.MustMarshalJSON(&metadata)
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:176]: I realize now (and this something that could be changed as well in the spec) that `metadata` doesn't change in this function, so we could simply just return `version` and save the marshalling.

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go [hunk 4]
```diff
@@ -471,3 +476,106 @@ func (suite *KeeperTestSuite) TestOnChanCloseConfirm() {
...
+		},
+		{
+			name: "failure: invalid metadata",
+			malleate: func() {
+				metadata.Address = TestOwnerAddress
+			},
+			expError: icatypes.ErrInvalidAccountAddress,
+		},
+		{
+			name: "failure: change connection id",
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:504]:
```suggestion
			name: "failure: change controller connection id",
```

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go [hunk 5]
```diff
@@ -471,3 +476,106 @@ func (suite *KeeperTestSuite) TestOnChanCloseConfirm() {
...
+		},
+		{
+			name: "failure: change host connection id",
+			malleate: func() {
+				metadata.HostConnectionId = differentConnectionID
+			},
+			expError: connectiontypes.ErrInvalidConnection,
+		},
+		{
+			name: "failure: invalid metadata string",
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:518]: Maybe this is more explicit:

```suggestion
			name: "failure: cannot decode version string",
```

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go [hunk 6]
```diff
@@ -471,3 +476,106 @@ func (suite *KeeperTestSuite) TestOnChanCloseConfirm() {
...
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {},
+			nil,
+		},
+		{
+			name: "failure: invalid metadata",
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:497]:
```suggestion
			name: "failure: change ICA address",
```

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go [hunk 7]
```diff
@@ -471,3 +476,106 @@ func (suite *KeeperTestSuite) TestOnChanCloseConfirm() {
...
+			path = NewICAPath(suite.chainA, suite.chainB)
+			suite.coordinator.SetupConnections(path)
+
+			err := SetupICAPath(path, TestOwnerAddress)
+			suite.Require().NoError(err)
+
+			channel := path.EndpointA.GetChannel()
+
+			currentMetadata, err := icatypes.ParseMedataFromString(channel.Version)
+			suite.Require().NoError(err)
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:550]: Maybe you could substitute this for `suite.chainA.GetSimApp().ICAControllerKeeper.GetAppMetadata(...)`.

**Review by crodriguezvega — COMMENTED** [review]
Thanks for working on the ICA upgrade callbacks, @chatton!

Besides the comments I left, I think we should also implement `OnChanUpgradeInit` in the host submodule and return an error saying that the upgrade handshake can only be initiated by the controller chain.

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 8]
```diff
@@ -136,3 +137,42 @@ func (Keeper) OnChanCloseConfirm(
...
+
+	channel, found := k.channelKeeper.GetChannel(ctx, portID, channelID)
+	if !found {
+		return "", errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "failed to retrieve channel %s on port %s", channelID, portID)
+	}
+
+	currentMetadata, err := icatypes.ParseMedataFromString(channel.Version)
+	if err != nil {
+		return "", err
+	}
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:160]: ah, it's a function on the host keeper, not the controller keeper

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go (hunk 3 above)
Comment (chatton) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:176]: nice catch!

**Comment by chatton** [comment]
> Besides the comments I left, I think we should also implement OnChanUpgradeInit in the host submodule and return an error saying that the upgrade handshake can only be initiated by the controller chain.

This is already being done [here](https://github.com/cosmos/ibc-go/blob/4383d623e41e811948ff106b3e9e58f6a4e4ebcb/modules/apps/27-interchain-accounts/host/ibc_module.go#L159-L162) is there somewhere else we need to error?

**Comment by chatton** [comment]
@crodriguezvega let me know if you're happy with the changes! If so I'll update the other handlers with the same changes 💪

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go]: Nice, I see that you implemented it also now in the contoller. 👍

**Review by crodriguezvega — APPROVED** [review]
Nice work, @chatton. Thanks for addressing all the comments!

**Review by damiannolan — APPROVED** [review]

**Review by crodriguezvega — COMMENTED** [review]
Great work, @chatton and @charleenfei!

**Pull request merged**